### PR TITLE
Updates based on testing

### PR DESCRIPTION
**Pull request by charles-cowart**
Updates based on testing on qiita-rc.

### CODE REVIEW

File: sequence_processing_pipeline/Pipeline.py [hunk 1]
```diff
@@ -347,11 +348,41 @@ def get_project_info(self):
...
-            return False
+        except PipelineError as e:
+            # we want to distinguish between a file that is clearly not a
+            # mapping file e.g. sample-file, and a mapping-file that is perhaps
+            # missing a column or has duplicate sample-names.
+            good_messages = ['duplicate sample-names detected:',
+                             'missing columns:']
+            for message in good_messages:
+                if str(e).startswith(message):
+                    return True
```
Comment (antgonza) [inline sequence_processing_pipeline/Pipeline.py:365]: Can this be just messages or explain why good vs. bad?

File: sequence_processing_pipeline/Pipeline.py [hunk 2]
```diff
@@ -347,11 +348,41 @@ def get_project_info(self):
...
+        Returns True if file follows basic sample-sheet format.
+        '''
+
+        # Previously SampleSheet() was used to test if file was a sample-
+        # sheet. SampleSheet() does not return an error on arbitrary-text
+        # files however. Checking to see if the file begins w/[Header]
+        # (ignoring any legacy comments) appears to be the best solution
+        # for now.
+        with open(sample_sheet_path, 'r') as f:
+            lines = f.readlines()
```
Comment (antgonza) [inline sequence_processing_pipeline/Pipeline.py:381]: This is reading all lines but only the first one is being used here, is this necessary?

File: sequence_processing_pipeline/Pipeline.py [hunk 3]
```diff
@@ -365,15 +396,35 @@ def _validate_mapping_file(mapping_file_path):
...
+                    return df
+            else:
+                if exp - set(df.columns) == exp:
+                    # since pandas.read_csv() will successfully open a sample-
+                    # sheet, if none of the expected columns exist then assume
+                    # that this file is not a mapping file.
+                    msg = "not a mapping-file"
+                else:
+                    msg = 'missing columns: %s' % ', '.join(
+                        exp - set(df.columns))
```
Comment (antgonza) [inline sequence_processing_pipeline/Pipeline.py:427]: Just noticed that this has an "interesting" logic, basically you `return df` in the middle of the function but you keep the msg and raise an error at the end. I normally have seen the opposite: you raise an error in the middle of the code and `return df` at the end (if no errors where found). The only time I have seen this, that I can remember, is when you can have multiple messages in a single run and you initialize the buffer `msgs = []` and on each new message you append to finally have a `if msgs: raise XX else: return XXX`, which is not the case here.

File: sequence_processing_pipeline/tests/test_Pipeline.py [hunk 4]
```diff
@@ -1009,11 +1010,30 @@ def test_mapping_file_validation(self):
...
             Pipeline(self.good_config_file, self.good_run_id,
-                     None, self.bad_mapping_file_path,
+                     None, self.mf_missing_column,
                      self.output_file_path,
                      self.qiita_id, None)
         self.assertEqual(str(e.exception), 'missing columns: tm50_8_tool')
 
+    def test_is_mapping_file(self):
+        # assert that a good sample-sheet is not a mapping-file
+        print(self.sample_sheet_path)
```
Comment (antgonza) [inline sequence_processing_pipeline/tests/test_Pipeline.py:1020]: rm?

**Review by antgonza — CHANGES_REQUESTED** [review]

File: sequence_processing_pipeline/Pipeline.py (hunk 1 above)
Comment (charles-cowart) [reply sequence_processing_pipeline/Pipeline.py:365]: messages sounds good to me.

File: sequence_processing_pipeline/Pipeline.py (hunk 2 above)
Comment (charles-cowart) [reply sequence_processing_pipeline/Pipeline.py:381]: sample-sheets are relatively small and it's a clean way to remove legacy comments at the top of the file but sure I can change it.

File: sequence_processing_pipeline/Pipeline.py (hunk 3 above)
Comment (charles-cowart) [reply sequence_processing_pipeline/Pipeline.py:427]: I think it's a clean way to capture external errors like ParserError as well as raise our own specific errors.

Comment (charles-cowart) [reply sequence_processing_pipeline/Pipeline.py:430]: Ty though, I will keep your comments in mind for the future.

File: sequence_processing_pipeline/Pipeline.py [hunk 5]
```diff
@@ -347,11 +348,43 @@ def get_project_info(self):
...
+        with open(sample_sheet_path, 'r') as f:
+            line = f.readline()
+            while line:
+                if line.startswith('#'):
+                    line = f.readline()
+                else:
+                    break
+
+            if line.startswith('[Header]'):
+                return True
```
Comment (antgonza) [inline sequence_processing_pipeline/Pipeline.py:385]: Should these lines be outside the `with open...`

File: sequence_processing_pipeline/Pipeline.py [hunk 6]
```diff
@@ -365,15 +398,35 @@ def _validate_mapping_file(mapping_file_path):
...
+            df.columns = df.columns.str.lower()
 
-            msg = 'missing columns: %s' % ', '.join(exp - set(df.columns))
+            # if the two sets of headers are equal, verify there are no
+            # duplicate sample-names and then return the dataframe and no
+            # error message.
+            if len(exp - set(df.columns)) == 0:
+                sample_names = df['sample_name'].tolist()
+                dupes = Counter(sample_names)
+                dupes = [x for x in dupes if dupes[x] > 1]
```
Comment (antgonza) [inline sequence_processing_pipeline/Pipeline.py:415]: Just realized that you could do this directly with pandas: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.value_counts.html

**Review by antgonza — CHANGES_REQUESTED** [review]

File: sequence_processing_pipeline/Pipeline.py (hunk 5 above)
Comment (charles-cowart) [reply sequence_processing_pipeline/Pipeline.py:385]: For this I would request to everything under the with open() as it makes it easy to process individual lines for potential comments and it allows us to not have to read the entire file. Having the entire while loop underneath the with open() allows us to do this.

File: sequence_processing_pipeline/Pipeline.py (hunk 6 above)
Comment (charles-cowart) [reply sequence_processing_pipeline/Pipeline.py]: nice! ty for the reference; that's much better!

**Review by antgonza — APPROVED** [review]

**Pull request merged**